design/spdk/spdk.md

```diff
@@ -0,0 +1,203 @@
+# Add SPDK to Rook
```

Comment:
@cyb70289 In my experience with CSI drivers, they can typically have a more static deployment and there is no need for an operator just for the csi driver. For example, have you considered the following?
- Create a helm chart to deploy the csi driver
- Add labels to the nodes where you want to deploy the spdk node daemons
- Add a node selector for the node daemons to only be deployed on those nodes
- Add annotations to each node with the node-specific settings such as the target type and port.
- The node daemon can load metadata about the annotations (it can start with a service account that allows access to the nodes)

If you can accomplish your goal with a helm chart, I would go with that instead of an operator and there would be no need to integrate with Rook.

Reply:
Thanks @travisn for the quick response!
I'm not familiar with helm. Will do some investigations.
Below are some requirements, would like to hear your advice. Thanks.
- deployment:
  1. deploy spdk statefulset on spdk nodes (helm ok)
  2. generate configmap including spdk pod IP (**is helm ok?**)
  3. deploy spdk-csi, mount the configmap in controller pod (helm ok)
  4. create storageclass, rbac, etc (helm ok)
- maintenance:
  - update configmap when storage nodes are added or removed (**is helm ok?**)

Reply:
> generate configmap including spdk pod IP (is helm ok?)

The pod IP can change every time the pod restarts. Even the operator can't detect the pod IP and make it reliable since the pod IP can change. If you need a constant endpoint, you need to create a K8s Service resource, which has a constant clusterIP. You could create the service with the helm chart since it's just another k8s resource.

You might also consider detecting the pod IP as part of the pod itself, either in the main container or with an init container if needed.

> update configmap when storage nodes are added or removed (is helm ok?)

Do you want to add a node IP to the configmap? That sounds different from the pod IP in the previous bullet. But if you have a pod run on that node (perhaps you have a daemonset that runs your daemons on all nodes with a certain label or other node selector), you can detect info of that node.

Reply:
Thanks for your advice.